src/projection.jl

```diff
@@ -52,93 +128,197 @@ function ProjectTo(x::T) where {T<:Union{<:Tuple,NamedTuple}}
 end
 
 # Generic
-(project::ProjectTo)(dx::AbstractThunk) = project(unthunk(dx))
 (::ProjectTo{T})(dx::T) where {T} = dx  # not always true, but we can special case for when it isn't
-(::ProjectTo{T})(dx::AbstractZero) where {T} = zero(T)
+(::ProjectTo{T})(dx::AbstractZero) where {T} = dx # zero(T)
+
+# Thunks
+(project::ProjectTo)(dx::Thunk) = Thunk(project ∘ dx.f)
+(project::ProjectTo)(dx::InplaceableThunk) = Thunk(project ∘ dx.val.f) # can't update in-place part
+(project::ProjectTo)(dx::AbstractThunk) = project(unthunk(dx))
 
-# Number
-ProjectTo(::T) where {T<:Number} = ProjectTo{T}()
+# Non-differentiable -- not so sure this is necessary. Keeping value like this is a bit awkward.
+for T in (:Bool, :Symbol, :Char, :String, :Val, :Type)
+    @eval ProjectTo(dx::$T) = ProjectTo{AbstractZero}(; value=dx)
+end
+(::ProjectTo{AbstractZero})(dx::AbstractZero) = dx
+(::ProjectTo{AbstractZero})(dx) = ZeroTangent()
+
+# Number -- presently preserves Float32, but not Int, is this ideal?
+ProjectTo(::T) where {T<:Number} = ProjectTo{float(T)}()
 (::ProjectTo{T})(dx::Number) where {T<:Number} = convert(T, dx)
 (::ProjectTo{T})(dx::Number) where {T<:Real} = convert(T, real(dx))
```

Comment:
This will mean `gradient(*, 2, 3) === (3.0, 2.0)` which seems fine. It will also mean Float32 is rigidly preserved, which will solve a lot of Flux performance bugs from accidental Float64 promotion.

What I worry a bit about is whether this closes the door on things like pushing Dual numbers through things. (Or more exotically, trying to use Unitful & having gradients with different units? That does not work at all now, maybe too exotic.)

One alternative would be to make `ProjectTo{Real}` and then allow any Real in the gradient.

Reply:
Maybe we need something like `Real32` which passes Float32 along, down-converts Float64, but also doesn't complain if it gets a dual number etc? Or maybe that's too much code? Not sure